### ISSUE TITLE
custom layer destroys mask when supports_masking = True

### ISSUE THREAD

**Issue by georgeMcMahon** [rstudio/keras#1225]
Hi,

I am writing a model which masks time dimensions and then applies a dense layer over each non-masked dimension. I am finding that mask is destroyed by my custom layer implementation. Rather I expect the normal behaviour of layers with an opt in around masking to pass the mask through the layer. Below is a minimum working example of my implementation in the function API that fails on a custom layer wrapper around dense and and example that works using the core dense layer. I can work around by recalling `tf$keras$layers$Masking` on the output of the custom layer but for small batches and a relu activation function, this doesn't exclude the possibility of accidental masking. 
```
    ##custom layer
MyDenseLayer <- R6::R6Class(
  "CustomLayer",
  inherit = KerasLayer,
  
  public = list(
    num_outputs = NULL,
    kernel = NULL,
    supports_masking = T,
    
    initialize = function(num_outputs) {
      self$num_outputs <- num_outputs
    },
    
    build = function(input_shape) {
      self$kernel <- self$add_weight(name = 'kernel',
                                     shape = list(input_shape[[2]], self$num_outputs))
    },
    
    call = function(x, mask = NULL) {
      # tf$matmul(x, self$kernel)
      mask
    },
    
    compute_mask = function(x, mask = NULL) {
      mask
    },
    
    compute_output_shape = function(input_shape) {
      list(input_shape[[1]], self$num_outputs)
    }
  )
)

layer_my_dense <- function(object, num_outputs,name = NULL, trainable = TRUE) {
  create_layer(MyDenseLayer, object, list(
    num_outputs = as.integer(num_outputs),
    name = name,
    trainable = trainable
  ))
}


  #fails
inputs <- tf$keras$layers$Input(shape=list(10L))
maskingLayer = tf$keras$layers$Masking(mask_value=-9)
masked_input = maskingLayer(inputs)
custom_layer = layer_my_dense(num_outputs=5L)
custom_layer_output = custom_layer(masked_input)
custom_layer$input_mask
# KerasTensor(type_spec=TensorSpec(shape=(None,), dtype=tf.bool, name=None), name='masking_103/Squeeze:0')
custom_layer$output_mask
# NULL

#works
inputs <- tf$keras$layers$Input(shape=list(10L))
maskingLayer = tf$keras$layers$Masking(mask_value=-9)
masked_input = maskingLayer(inputs)
dense_layer = tf$keras$layers$Dense(5L)
dense_layer_output = dense_layer(masked_input)
dense_layer$input_mask
# KerasTensor(type_spec=TensorSpec(shape=(None,), dtype=tf.bool, name=None), name='masking_103/Squeeze:0')
dense_layer$output_mask
# KerasTensor(type_spec=TensorSpec(shape=(None,), dtype=tf.bool, name=None), name='Placeholder_1:0')
```

**Comment by georgeMcMahon** [comment]
apologies here is my sessionInfo

`
R version 4.0.5 (2021-03-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] datasets  grDevices stats     utils     graphics  methods   base     

other attached packages:
 [1] reticulate_1.18              tfdatasets_2.2.0            tensorflow_2.4.0         
[3] keras_2.3.0.0

**Comment by t-kalinowski** [comment]
Hi, Thanks for filing. You are correct, custom layers currently do not have support for a user-defined `compute_mask` method. Please stay tuned, a fix is in the works.

**Comment by t-kalinowski** [comment]
Hi, thanks for filing. This should work now in the dev version of keras.

Note, you'll have to make some minor changes to your example above to make it work, notably, you will want to:

* change `inherit = KerasLayer` to `inherit = keras$layers$Layer` 
* change `initialize()` to call `super`: 
```R
     initialize = function(num_outputs, ...) {
        super$initialize(...)
        self$num_outputs <- num_outputs
      }
```
I've included a modified version of your example in the unit tests, please take a look if that's helpful as a usage guide: https://github.com/rstudio/keras/blob/86da954238516802039fb9c789d8fb93a21d032e/tests/testthat/test-custom-layers.R#L177
